pactflow/client/utils.ts

```diff
@@ -0,0 +1,101 @@
+import {
+  RemoteOpenAPIDocument,
+  RemoteOpenAPIDocumentSchema,
+  OpenAPI,
+} from "./ai.js";
+import yaml from "js-yaml";
+// @ts-expect-error missing type declarations
+import Swagger from "swagger-client";
+
+/**
+ * Resolve the OpenAPI specification from the provided input.
+ *
+ * @param remoteOpenAPIDocument The remote OpenAPI document to resolve.
+ * @returns The resolved OpenAPI document.
+ * @throws Error if the resolution fails.
+ */
+export async function resolveOpenAPISpec(
+  remoteOpenAPIDocument: RemoteOpenAPIDocument
+): Promise<OpenAPI> {
+  const openAPISchema = RemoteOpenAPIDocumentSchema.safeParse(
+    remoteOpenAPIDocument
+  );
+  if (openAPISchema.error || !remoteOpenAPIDocument) {
+    throw new Error(
+      `Invalid RemoteOpenAPIDocument: ${JSON.stringify(
+        openAPISchema.error?.issues
+      )}`
+    );
+  }
+
+  const unresolvedSpec = await getRemoteSpecContents(openAPISchema.data);
+  const resolvedSpec = await Swagger.resolve({ spec: unresolvedSpec });
+
+  if (resolvedSpec.errors?.length) {
+    throw new Error(
+      `Failed to resolve OpenAPI document: ${resolvedSpec.errors?.join(", ")}`
+    );
+  }
+
+  return resolvedSpec.spec;
+}
+
+/**
+ * Fetch the contents of a remote OpenAPI document.
+ *
+ * @param openAPISchema The schema for the remote OpenAPI document.
+ * @returns A promise that resolves to a map of the OpenAPI document contents.
+ * @throws Error if the URL is not provided or the fetch fails.
+ */
+export async function getRemoteSpecContents(
+  openAPISchema: RemoteOpenAPIDocument
+): Promise<any> {
+  if (!openAPISchema.url) {
+    throw new Error("'url' must be provided.");
+  }
+
+  let headers = {};
+  if (openAPISchema.authToken) {
+    headers = {
+      Authorization: `${openAPISchema.authScheme ?? "Bearer"} ${
+        openAPISchema.authToken
+      }`,
+    };
+  }
+
+  const remoteSpec = await fetch(openAPISchema.url, {
+    headers,
+    method: "GET",
+  });
+
+  const specRawBody = await remoteSpec.text();
+
+  try {
+    return JSON.parse(specRawBody);
+  } catch {
+    try {
+      return JSON.parse(JSON.stringify(yaml.load(specRawBody), null, 2));
```

Comment:
It feels like you're just doing $f^{-1}(f(x)) = x$ here (i.e., `stringify` and `parse` are inverses of each other, and therefore their composition results in the identity function $I : x \mapsto x$).

Why couldn't you have:


```suggestion
      return yaml.load(specRawBody);
```

Reply:
The Swagger-js lib can only parse a JS object, which is why even the yaml needs to be converted into a json and returned back.